internal/controller/dragonfly_controller.go

```diff
@@ -88,7 +89,127 @@ func (r *DragonflyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 			return ctrl.Result{}, err
 		}
 
-		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Created", "Created resources for Dragonfly object")
+		r.EventRecorder.Event(&df, corev1.EventTypeNormal, "Resources", "Created resources")
+		return ctrl.Result{}, nil
+	} else if df.Status.Phase == PhaseRollout {
+		// This is a Rollout
+		log.Info("Rolling out new version")
+		var updatedStatefulset appsv1.StatefulSet
+		if err := r.Get(ctx, client.ObjectKey{Namespace: df.Namespace, Name: df.Name}, &updatedStatefulset); err != nil {
+			log.Error(err, "could not get statefulset")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		// get pods of the statefulset
+		var pods corev1.PodList
+		if err := r.List(ctx, &pods, client.InNamespace(df.Namespace), client.MatchingLabels(map[string]string{
+			"app":                               df.Name,
+			resources.KubernetesAppNameLabelKey: "dragonfly",
+		})); err != nil {
+			log.Error(err, "could not list pods")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		if len(pods.Items) != int(*updatedStatefulset.Spec.Replicas) {
+			log.Info("Waiting for all replicas to be ready")
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		// filter replicas
+		var master corev1.Pod
+		replicas := make([]corev1.Pod, 0)
+		for _, pod := range pods.Items {
+			if _, ok := pod.Labels[resources.Role]; ok {
+				if pod.Labels[resources.Role] == resources.Replica {
+					replicas = append(replicas, pod)
+				} else if pod.Labels[resources.Role] == resources.Master {
+					master = pod
+				}
+			} else {
+				log.Info("found pod without label", "pod", pod.Name)
+				// retry after they are ready
+				return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+			}
+		}
+
+		// Check if all the latest replicas had a full sync
```

Comment:
// We want to update the replicas first then the master
// We want to have at most one updated replica in full sync phase at a time